063-topic-replication-factor-change.md

```diff
@@ -0,0 +1,271 @@
+
+### KafkaTopic CRD changes
+
+During the whole replicas change process the KafkaTopic remains in a Ready state.
+A replicas change is executed asynchronously, and takes more than one reconciliation to complete.
+
+A new optional KafkaTopicStatus subsection called `replicasChange` is used to update the user, and keep track of the replicas change state.
+At any given time, the replicas change can be pending, ongoing, or completed.
+
+- **Pending**: Not in Cruise Control's task queue (not yet sent or request error).
+  ```sh
+  status:
+    conditions:
+    - lastTransitionTime: "2024-01-18T16:13:50.490918232Z"
+      status: "True"
+      type: Ready
+    observedGeneration: 2
+    replicasChange:
+      completed: false
+    topicName: my-topic
+  ```
+
+- **Ongoing**: In Cruise Control's task queue, but execution not started, or not completed.
+  ```sh
+  status:
+    conditions:
+    - lastTransitionTime: "2024-01-18T16:13:53.490918232Z"
+      status: "True"
+      type: Ready
+    observedGeneration: 3
+    replicasChange:
+      completed: false
+      userTaskId: 1aa418ca-53ed-4b93-b0a4-58413c4fc0cb
+    topicName: my-topic
+  ```
+
+- **Completed**: Cruise Control's task execution completed (target replication factor reconciled).
+  ```sh
+  status:
+    conditions:
+    - lastTransitionTime: "2024-01-18T16:13:58.490918232Z"
+      status: "True"
+      type: Ready
+    observedGeneration: 4
+    topicName: my-topic
+  ```
+
+### Reconciliation logic
+
+At runtime, the Topic Operator watches and periodically reconciles the configuration of all managed and unpaused KafkaTopic resources.
+Among the other configurations, the reconciliation logic detects `KafkaTopic.spec.replicas` changes by comparing with the topic replication factor value in Kafka.
+
+A ReplicasChangeClient is used to isolate the logic required to call the Cruise Control REST API from the rest of the reconciliation logic.
+The ReplicasChangeClient supports sending multiple replicas changes at the same time.
+
+On every reconciliation loop, when Cruise Control integration is enabled and replicas changes are detected, the BatchingTopicController uses the ReplicasChangeClient for the following operations:
+
+1. Request pending replicas changes (uses `topic_configuration` endpoint, and returns topics with pending and ongoing changes).
+2. Check ongoing replicas changes (uses `user_tasks` endpoint, and returns topics with completed changes).
+
+The result of these operations is used to update the KafkaTopic status.
+
+### Error handling
+
+When there is an error, the replicas change remains pending or ongoing (no status update), and it is indefinitely retried by the periodic reconciliation.
+All errors are written in the Topic Operator logs, so that the user can verify the reason of a never ending replicas change.
```

Comment:
Should/can the errors be added also to the conditions as warnings? I'm not sure if they are persistent, so not sure if it is really possible or if they will be added and immediately removed in the next reconciliation few milliseconds later.

Reply:
Added a message field to `replicasChange`, that we can use for any message to the user. For the moment I'll use it only for errors/warnings.